[GFC] Initial implementation of grid item placement.

sammygill opened the pull request:
#### 5063ef34f4f66ebb467473eb9ed62e5a40cdbe7b
<pre>
[GFC] Initial implementation of grid item placement.
<a href="https://bugs.webkit.org/show_bug.cgi?id=298989">https://bugs.webkit.org/show_bug.cgi?id=298989</a>
<a href="https://rdar.apple.com/problem/160729099">rdar://problem/160729099</a>

Reviewed by Sam Weinig and Alan Baradlay.

This patch introduces the initial implementation of grid item placement
by implementing some overall logic around simple content. In this case,
simple content refers to grid content in which all items are explicitly
placed, span only a single track, and other placement constraints which
should be documented through the code. In addition, we only support
placement by placing items within columns and grids that are explicitly
defined via the grid-template-{columns, rows} properties.

The overall purpose of this patch is not to implement all of the logic
required to satisfy item placement in all scenarios but rather provide
some overall scaffolding in which we can build upon one at a time. We
liberally use ASSERT_NOT_IMPLEMENTED to help identify these areas in which
we can and will need to build upon in future work.

* Source/WebCore/layout/formattingContexts/grid/GridFormattingContext.cpp:
(WebCore::Layout::GridFormattingContext::constructUnplacedGridItems const):
* Source/WebCore/layout/formattingContexts/grid/GridFormattingContext.h:
(WebCore::Layout::GridFormattingContext::root const):
* Source/WebCore/layout/formattingContexts/grid/GridLayout.cpp:
(WebCore::Layout::GridLayout::layout):
(WebCore::Layout::GridLayout::placeGridItems):
* Source/WebCore/layout/formattingContexts/grid/GridLayout.h:
(WebCore::Layout::GridLayout::gridContainer const):
(WebCore::Layout::GridLayout::gridContainerStyle const):

* Source/WebCore/layout/formattingContexts/grid/ImplicitGrid.cpp: Added.
(WebCore::Layout::ImplicitGrid::ImplicitGrid):
The main purpose of ImplicitGrid is to serve as a helper structure to
keep track of some state during item placement. Since an implicit grid
is constructed from the explicit grid, as described in the spec, the
idea is that the constructor takes in the number of explicit columns and
rows that are specified in the style. For now, we only take in the
columns and rows that are specified by grid-template-{columns, rows}, but
eventually, we will also need to take into consideration the fact
that the number of explicit tracks can also be influenced by the
grid-template-areas property.

(WebCore::Layout::ImplicitGrid::insertUnplacedGridItem):
Users of ImplicitGrid will insert unplaced items through this API, and
the ImplicitGrid will handle resizing the grid (i.e. adding implicit
columns and rows) as necessary to fit the items. This logic isn&apos;t
implemented currently since we are only worrying about items that fit
within the explicitly sized grid, and as a result, we should not need to
expect to resize the grid.

* Source/WebCore/layout/formattingContexts/grid/PlacedGridItem.cpp:
(WebCore::Layout::PlacedGridItem::PlacedGridItem):
* Source/WebCore/layout/formattingContexts/grid/PlacedGridItem.h:
A PlacedGridItem is the result of resolving the placement (i.e. the grid
area) of an UnplacedGridItem. This structure is very similar to
UnplacedGridItem, with the main difference being that the positions are
typed by size_t wthin a GridAreaLines struct to indicate the fact that
the location is resolved within the implicit grid.

(WebCore::Layout::UnplacedGridItem::explicitColumnStart const):
(WebCore::Layout::UnplacedGridItem::explicitColumnEnd const):
(WebCore::Layout::UnplacedGridItem::explicitRowStart const):
(WebCore::Layout::UnplacedGridItem::explicitRowEnd const):
Some straightforward helper functions to access the item&apos;s explicit
column/row positions. Since the grid lines in the properties&apos; grammar are
1-indexed, we also subtract 1 from these values so that they map nicely
to our 0 indexed structures. For example, grid-column-start: 1 and
grid-column-end: 2 would return [0, 1] which would map the item to be
placed into GridMatrix[rowIndex][0]. We also leave negative line numbers
to be handled in a future patch to maintain simplicity.

* Source/WebCore/layout/formattingContexts/grid/UnplacedGridItem.h:
In order to place an UnplacedGridItem we need to know its placement
(position and span) which are given from the grid-{column, row}-{start, end}
properties. Any information that is needed to resolve a grid item&apos;s
placement should be stored on this class and be accessed through a
dedicated API, such as the explicit position helper functions above.

Also, we change PlacedGridItems from being a simple Vector&lt;PlacedGridItem&gt;
to a struct instead. This is because it will organize the different
unplaced grid items into a way that is easily accessible to the
placement logic in a manner that is similar to how the spec reads. For
example, the first step in placement is to place all non-auto positioned
items so this will be a collection of items in UnplacedGridItems that
the code can get to perform that step.

Canonical link: <a href="https://commits.webkit.org/300141@main">https://commits.webkit.org/300141@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cd4f215e29d31aa460c91fe3bfa69d9b4caf7e06

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/121366 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/41063 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/31722 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/127806 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/73449 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/95ed812b-dbcb-4ddb-ba4a-0781a487c315) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/123242 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/41765 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/49642 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/92189 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/61340 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/840e987a-52ea-4af7-b7c6-1db3bfd8d040) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/124318 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/33351 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/108744 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/72865 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/a0e6e155-69c9-4860-b309-a229e624584d) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/32368 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/26883 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/71385 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/102845 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/27058 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/130640 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/48294 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/36718 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/100784 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/48662 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/104943 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/100689 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25544 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/46099 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/24164 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/44990 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/48152 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/53865 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/47624 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/50970 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/49306 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/4bab307e416fe708be4c200bc27aa2941eb77403. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50831

weinig commented on Source/WebCore/layout/formattingContexts/grid/GridLayout.cpp:
```diff
@@ -25,6 +25,10 @@
 
 #include "config.h"
 #include "GridLayout.h"
+#include "ImplicitGrid.h"
```
Missing newline after `#include "GridLayout.h"`

weinig commented on Source/WebCore/layout/formattingContexts/grid/GridFormattingContext.h:
```diff
@@ -50,6 +50,8 @@ class GridFormattingContext : public CanMakeCheckedPtr<GridFormattingContext> {
 
     void layout(GridLayoutConstraints);
+
+    const ElementBox& root() const { return m_gridBox; }
```
Missing newline before `private:`

sammygill replied:
Fixed

weinig commented on Source/WebCore/layout/formattingContexts/grid/GridLayout.cpp:
```diff
@@ -33,6 +37,28 @@ GridLayout::GridLayout(const GridFormattingContext& gridFormattingContext)
+    UNUSED_VARIABLE(placedGridItems);
+}
+
+// 8.5. Grid Item Placement Algorithm.
```
Consider adding a spec link here.

sammygill replied:
Done

weinig commented on Source/WebCore/layout/formattingContexts/grid/ImplicitGrid.cpp:
```diff
@@ -0,0 +1,119 @@
+        return;
+    }
+
+    auto columnsCount = static_cast<int>(this->columnsCount());
```
What is the intention of this cast?

weinig commented on Source/WebCore/layout/formattingContexts/grid/ImplicitGrid.h:
```diff
@@ -0,0 +1,55 @@
+
+#pragma once
+
+#include "LayoutElementBox.h"
```
Is this used?

sammygill replied:
Nope removed

weinig commented on Source/WebCore/layout/formattingContexts/grid/ImplicitGrid.cpp:
```diff
@@ -0,0 +1,119 @@
+#include "PlacedGridItem.h"
+#include "UnplacedGridItem.h"
+
+#include <wtf/Range.h>
```
Extra newline.

sammygill replied:
Fixed

weinig commented on Source/WebCore/layout/formattingContexts/grid/GridLayout.h:
```diff
@@ -26,19 +26,27 @@
+    static PlacedGridItems placeGridItems(const UnplacedGridItems&, size_t gridTemplateColumnsTracksCount, size_t gridTemplateRowsTracksCount);
+
+    const ElementBox& gridContainer() const { return m_gridFormattingContext->root(); }
+    const RenderStyle& gridContainerStyle() const { return gridContainer().style(); }
```
It would let you remove the #include of "LayoutElementBox.h" here as well.

sammygill replied:
Fair enough. Moved

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/08a19dc98ef000ed73f45efde4c56e5048b9d824. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50831

sammygill commented on Source/WebCore/layout/formattingContexts/grid/PlacedGridItem.h:
```diff
@@ -0,0 +1,48 @@
+private:
+    const CheckedRef<const ElementBox> m_layoutBox;
+
+    std::pair<size_t, size_t> m_columnPositon;
```
Good idea! I added a struct for the lines defined by this grid area so I think this should make things clearer

alanbaradlay reviewed (approved):
It would be great to have more spec text or reference to steps (sections) in some of these functions.

sammygill commented:
Thanks! I've added some more comments so hopefully they are helpful! I also had to add some helpers to `PlacedGridItem` in order to appease 

```
C:\BW\Win-Build-EWS\build\Source\WebCore\layout/formattingContexts/grid\PlacedGridItem.h(50,19): error: private field 'm_gridAreaLines' is not used [-Werror,-Wunused-private-field]
   50 |     GridAreaLines m_gridAreaLines;
```

sammygill committed:
[GFC] Initial implementation of grid item placement.
https://bugs.webkit.org/show_bug.cgi?id=298989
rdar://problem/160729099

Reviewed by Sam Weinig and Alan Baradlay.

This patch introduces the initial implementation of grid item placement
by implementing some overall logic around simple content. In this case,
simple content refers to grid content in which all items are explicitly
placed, span only a single track, and other placement constraints which
should be documented through the code. In addition, we only support
placement by placing items within columns and grids that are explicitly
defined via the grid-template-{columns, rows} properties.

The overall purpose of this patch is not to implement all of the logic
required to satisfy item placement in all scenarios but rather provide
some overall scaffolding in which we can build upon one at a time. We
liberally use ASSERT_NOT_IMPLEMENTED to help identify these areas in which
we can and will need to build upon in future work.

* Source/WebCore/layout/formattingContexts/grid/GridFormattingContext.cpp:
(WebCore::Layout::GridFormattingContext::constructUnplacedGridItems const):
* Source/WebCore/layout/formattingContexts/grid/GridFormattingContext.h:
(WebCore::Layout::GridFormattingContext::root const):
* Source/WebCore/layout/formattingContexts/grid/GridLayout.cpp:
(WebCore::Layout::GridLayout::layout):
(WebCore::Layout::GridLayout::placeGridItems):
* Source/WebCore/layout/formattingContexts/grid/GridLayout.h:
(WebCore::Layout::GridLayout::gridContainer const):
(WebCore::Layout::GridLayout::gridContainerStyle const):

* Source/WebCore/layout/formattingContexts/grid/ImplicitGrid.cpp: Added.
(WebCore::Layout::ImplicitGrid::ImplicitGrid):
The main purpose of ImplicitGrid is to serve as a helper structure to
keep track of some state during item placement. Since an implicit grid
is constructed from the explicit grid, as described in the spec, the
idea is that the constructor takes in the number of explicit columns and
rows that are specified in the style. For now, we only take in the
columns and rows that are specified by grid-template-{columns, rows}, but
eventually, we will also need to take into consideration the fact
that the number of explicit tracks can also be influenced by the
grid-template-areas property.

(WebCore::Layout::ImplicitGrid::insertUnplacedGridItem):
Users of ImplicitGrid will insert unplaced items through this API, and
the ImplicitGrid will handle resizing the grid (i.e. adding implicit
columns and rows) as necessary to fit the items. This logic isn't
implemented currently since we are only worrying about items that fit
within the explicitly sized grid, and as a result, we should not need to
expect to resize the grid.

* Source/WebCore/layout/formattingContexts/grid/PlacedGridItem.cpp:
(WebCore::Layout::PlacedGridItem::PlacedGridItem):
* Source/WebCore/layout/formattingContexts/grid/PlacedGridItem.h:
A PlacedGridItem is the result of resolving the placement (i.e. the grid
area) of an UnplacedGridItem. This structure is very similar to
UnplacedGridItem, with the main difference being that the positions are
typed by size_t wthin a GridAreaLines struct to indicate the fact that
the location is resolved within the implicit grid.

(WebCore::Layout::UnplacedGridItem::explicitColumnStart const):
(WebCore::Layout::UnplacedGridItem::explicitColumnEnd const):
(WebCore::Layout::UnplacedGridItem::explicitRowStart const):
(WebCore::Layout::UnplacedGridItem::explicitRowEnd const):
Some straightforward helper functions to access the item's explicit
column/row positions. Since the grid lines in the properties' grammar are
1-indexed, we also subtract 1 from these values so that they map nicely
to our 0 indexed structures. For example, grid-column-start: 1 and
grid-column-end: 2 would return [0, 1] which would map the item to be
placed into GridMatrix[rowIndex][0]. We also leave negative line numbers
to be handled in a future patch to maintain simplicity.

* Source/WebCore/layout/formattingContexts/grid/UnplacedGridItem.h:
In order to place an UnplacedGridItem we need to know its placement
(position and span) which are given from the grid-{column, row}-{start, end}
properties. Any information that is needed to resolve a grid item's
placement should be stored on this class and be accessed through a
dedicated API, such as the explicit position helper functions above.

Also, we change PlacedGridItems from being a simple Vector<PlacedGridItem>
to a struct instead. This is because it will organize the different
unplaced grid items into a way that is easily accessible to the
placement logic in a manner that is similar to how the spec reads. For
example, the first step in placement is to place all non-auto positioned
items so this will be a collection of items in UnplacedGridItems that
the code can get to perform that step.

Canonical link: https://commits.webkit.org/300141@main
